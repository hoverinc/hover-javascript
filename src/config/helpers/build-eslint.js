const {
  rules: airbnbRules,
} = require('eslint-config-airbnb-typescript/lib/shared')

const {hasAnyDep} = require('../../utils')
const {testMatch} = require('../jest.config')

const withBaseConfig = base => variant =>
  require.resolve(base + (variant ? `/${variant}` : ''))

const airbnb = withBaseConfig('eslint-config-airbnb-typescript')
const prettier = withBaseConfig('eslint-config-prettier')

const hasReact = hasAnyDep('react')

const parserRules = (typescript = false, react = false) => {
  const isOff = off => (off ? 'off' : 'error')

  const propTypes = react ? {'react/prop-types': isOff(typescript)} : {}

  return {
    'no-implied-eval': isOff(typescript),
    'no-throw-literal': isOff(typescript),
    '@typescript-eslint/no-implied-eval': isOff(!typescript),
    '@typescript-eslint/no-throw-literal': isOff(!typescript),
    ...propTypes,
  }
}

const buildConfig = ({withReact = false} = {}) => {
  const isReact = withReact || hasReact
  const ifReact = (t, f) => (isReact ? t : f)

  return {
    plugins: ['prettier', 'jest', ifReact('react-hooks')].filter(Boolean),
    extends: [
      ifReact(airbnb(), airbnb('base')),
      'plugin:jest/recommended',
      prettier(),
      prettier('@typescript-eslint'),
      ifReact(prettier('react')),
      ifReact('plugin:react-hooks/recommended'),
    ].filter(Boolean),
    rules: {
      'import/prefer-default-export': 'off',
      'import/no-extraneous-dependencies': [
        'error',
        {
          devDependencies: airbnbRules[
            'import/no-extraneous-dependencies'
          ][1].devDependencies.concat([
            'jest/**',
            'e2e/**',
            '**/*.config.{js,ts}',
          ]),
          optionalDependencies: false,
        },
      ],
      'import/order': [
        'error',
        {
          alphabetize: {order: 'asc'},
          'newlines-between': 'always',
          pathGroups: [
            {pattern: 'src/**/*', group: 'parent', position: 'before'},
            {pattern: 'assets/**/*', group: 'parent', position: 'before'},
          ],
          pathGroupsExcludedImportTypes: ['builtin'],
        },
      ],
      'no-void': ['error', {allowAsStatement: true}],
      'prettier/prettier': 'error',
      'sort-imports': ['error', {ignoreDeclarationSort: true}],
      ...parserRules(false, isReact),
    },
    overrides: [
      {
        files: ['**/*.ts?(x)'],
        extends: [
          'plugin:@typescript-eslint/recommended-requiring-type-checking',
        ],
        rules: {
          ...parserRules(true, isReact),
        },
      },
      {
        files: testMatch,
        rules: {
          'no-empty': ['error', {allowEmptyCatch: true}],
        },
      },
      {
        files: ['**/*/__tests__/helpers/**/*', '**/*/__tests__/utils/**/*'],
        rules: {
          'jest/no-export': 'off',
        },
      },
    ],
  }
}

module.exports = {buildConfig}
