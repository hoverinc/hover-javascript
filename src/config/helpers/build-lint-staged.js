const {resolveHoverScripts, resolveBin} = require('../../utils')

const hoverScripts = resolveHoverScripts()
const doctoc = resolveBin('doctoc')

const defaultTestCommand = `${hoverScripts} test --findRelatedTests --passWithNoTests`

const sourceExtensions = ['js', 'jsx', 'ts', 'tsx']

const readmeGlob = 'README.md'
const formatGlob =
  '*.+(js|jsx|json|json5|yml|yaml|css|less|scss|ts|tsx|md|graphql|mdx|vue)'

// This works around the limitation imposed by using globs as keys in the
// configuration. We want to run the lint and test commands on the same glob,
// but we want to do so in parallel. If we specify the commands in an array
// under the same glob key, they always run in sequence.
const lintGlob = `*.+(${sourceExtensions.join('|')})`
const testGlob = `*.+(${sourceExtensions.reverse().join('|')})`

const buildConfig = (toc = true, testCommand = defaultTestCommand) => ({
  ...(toc ? {[readmeGlob]: [`${doctoc} --maxlevel 4 --notitle`]} : {}),
  [formatGlob]: [`${hoverScripts} format`],
  [lintGlob]: [`${hoverScripts} lint`],
  [testGlob]: [testCommand],
})

module.exports = {buildConfig}
