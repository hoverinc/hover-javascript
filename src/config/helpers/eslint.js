const {rules} = require('eslint-config-airbnb-typescript/lib/shared')

const {hasAnyDep} = require('../../utils')
const {testMatch} = require('../jest.config')

const withBaseConfig = base => variant =>
  require.resolve(base + (variant ? `/${variant}` : ''))

const airbnb = withBaseConfig('eslint-config-airbnb-typescript')
const prettier = withBaseConfig('eslint-config-prettier')

const hasReact = hasAnyDep('react')

const parserRules = (typescript = false) => {
  const isOff = off => (off ? 'off' : 'error')

  return {
    'no-implied-eval': isOff(typescript),
    'no-throw-literal': isOff(typescript),
    '@typescript-eslint/no-implied-eval': isOff(!typescript),
    '@typescript-eslint/no-throw-literal': isOff(!typescript),
  }
}

const buildConfig = ({withReact = false} = {}) => {
  const ifReact = (t, f) => (withReact || hasReact ? t : f)

  return {
    plugins: ['prettier', 'jest', ifReact('react-hooks')].filter(Boolean),
    extends: [
      ifReact(airbnb(), airbnb('base')),
      'plugin:jest/recommended',
      prettier(),
      prettier('@typescript-eslint'),
      ifReact(prettier('react')),
    ].filter(Boolean),
    rules: {
      'prettier/prettier': 'error',
      'import/prefer-default-export': 'off',
      'import/no-extraneous-dependencies': [
        'error',
        {
          devDependencies: rules[
            'import/no-extraneous-dependencies'
          ][1].devDependencies.concat([
            'jest/**',
            'e2e/**',
            '**/prettier.config.js',
          ]),
          optionalDependencies: false,
        },
      ],
      ...parserRules(),
    },
    overrides: [
      {
        files: ['**/*.ts?(x)'],
        parserOptions: {
          project: './tsconfig.json',
        },
        extends: [
          'plugin:@typescript-eslint/recommended-requiring-type-checking',
        ],
        rules: {
          ...parserRules(true),
        },
      },
      {
        files: testMatch,
        rules: {
          'no-empty': ['error', {allowEmptyCatch: true}],
        },
      },
    ],
  }
}

module.exports = {buildConfig}
