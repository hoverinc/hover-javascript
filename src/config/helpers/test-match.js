const toGlob = extensions => `*.+(${extensions.join('|')})`

const testMatchExtensions = ['js', 'jsx', 'ts', 'tsx']
const testMatchGlob = toGlob(testMatchExtensions)
const testMatchSuffixGlob = toGlob(
  testMatchExtensions.map(extension => 'test.'.concat(extension)),
)

const testMatch = [
  `**/__tests__/**/${testMatchGlob}`,
  `test/**/${testMatchGlob}`,
  `test/**/${testMatchSuffixGlob}`,
  `e2e/**/${testMatchSuffixGlob}`,
  `**/${testMatchSuffixGlob}`,
]

module.exports = {
  testMatchExtensions,
  testMatchGlob,
  testMatchSuffixGlob,
  testMatch,
}
