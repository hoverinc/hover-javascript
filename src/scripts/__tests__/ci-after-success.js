import cases from 'jest-in-case';
import { unquoteSerializer } from './helpers/serializers';

expect.addSnapshotSerializer(unquoteSerializer);

cases(
  'ci-after-success',
  ({
    version = '0.0.0-semantically-released',
    hasCoverageDir = true,
    isOptedOutOfCoverage = false,
    env = {
      CI: 'true',
      CF_BRANCH: 'master',
    },
  }) => {
    // beforeEach
    const { sync: crossSpawnSyncMock } = require('cross-spawn');
    const utils = require('../../utils');
    utils.resolveBin = (modName, { executable = modName } = {}) => executable;
    const originalEnvs = Object.keys(env).map(envKey => {
      const orig = process.env[envKey];
      process.env[envKey] = env[envKey];
      return orig;
    });
    const originalLog = console.log;
    const originalExit = process.exit;
    process.exit = jest.fn();
    console.log = jest.fn();

    // tests
    if (version) {
      utils.pkg.version = version;
    }
    utils.hasFile = () => hasCoverageDir;
    process.env.SKIP_CODECOV = isOptedOutOfCoverage;
    require('../ci-after-success');

    expect(console.log.mock.calls).toMatchSnapshot();
    const commands = crossSpawnSyncMock.mock.calls.map(
      call => `${call[0]} ${call[1].join(' ')}`,
    );
    expect(commands).toMatchSnapshot();

    // afterEach
    process.exit = originalExit;
    console.log = originalLog;
    Object.keys(originalEnvs).forEach(envKey => {
      process.env[envKey] = env[envKey];
    });
    jest.resetModules();
  },
  {
    'calls concurrently with both scripts when on ci': {},
    'calls concurrently with both scripts when on travis': {
      env: {
        TRAVIS: 'true',
        TRAVIS_BRANCH: 'master',
        TRAVIS_PULL_REQUEST: 'false',
      },
    },
    'runs autorelease script on alternate release branch "next"': {
      env: {
        CI: 'true',
        CF_BRANCH: 'next',
        TRAVIS_PULL_REQUEST: 'false',
      },
    },
    'does not do the autorelease script when the version is different': {
      version: '1.2.3',
    },
    'does not do the codecov script when there is no coverage directory': {
      hasCoverageDir: false,
    },
    'does not do the codecov script when opted out': {
      isOptedOutOfCoverage: true,
    },
    'does not do autorelease script when running on travis but in a pull request': {
      env: {
        TRAVIS: 'true',
        TRAVIS_BRANCH: 'master',
        TRAVIS_PULL_REQUEST: 'true',
      },
    },
    'does not run either script when no coverage dir and not the right version': {
      hasCoverageDir: false,
      version: '1.2.3',
    },
  },
);
