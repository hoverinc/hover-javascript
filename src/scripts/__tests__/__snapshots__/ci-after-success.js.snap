// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ci-after-success calls concurrently with both scripts when on ci 1`] = `Array []`;

exports[`ci-after-success calls concurrently with both scripts when on ci 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov,release --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'" "echo installing semantic-release && npx -p semantic-release@17 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`ci-after-success calls concurrently with both scripts when on travis 1`] = `Array []`;

exports[`ci-after-success calls concurrently with both scripts when on travis 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov,release --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'" "echo installing semantic-release && npx -p semantic-release@17 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`ci-after-success does not do autorelease script when running on travis but in a pull request 1`] = `Array []`;

exports[`ci-after-success does not do autorelease script when running on travis but in a pull request 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov --prefix-colors bgBlue.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'",
]
`;

exports[`ci-after-success does not do the autorelease script when the version is different 1`] = `Array []`;

exports[`ci-after-success does not do the autorelease script when the version is different 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov --prefix-colors bgBlue.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'",
]
`;

exports[`ci-after-success does not do the codecov script when opted out 1`] = `Array []`;

exports[`ci-after-success does not do the codecov script when opted out 2`] = `
Array [
  concurrently --prefix [{name}] --names release --prefix-colors bgBlue.bold.reset "echo installing semantic-release && npx -p semantic-release@17 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`ci-after-success does not do the codecov script when there is no coverage directory 1`] = `Array []`;

exports[`ci-after-success does not do the codecov script when there is no coverage directory 2`] = `
Array [
  concurrently --prefix [{name}] --names release --prefix-colors bgBlue.bold.reset "echo installing semantic-release && npx -p semantic-release@17 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`ci-after-success does not run either script when no coverage dir and not the right version 1`] = `
Array [
  Array [
    No need to autorelease or report coverage. Skipping ci-after-success script...,
  ],
]
`;

exports[`ci-after-success does not run either script when no coverage dir and not the right version 2`] = `Array []`;

exports[`ci-after-success runs autorelease script on alternate release branch "next" 1`] = `Array []`;

exports[`ci-after-success runs autorelease script on alternate release branch "next" 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov,release --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'" "echo installing semantic-release && npx -p semantic-release@17 -c 'echo running semantic-release && semantic-release'",
]
`;
